hermes/tns.py

```diff
@@ -0,0 +1,241 @@
+                'exptime': str(earliest_nondetection.get('exposure_time', '')),
+                'observer': earliest_nondetection.get('observer', ''),
+                'comments': earliest_nondetection.get('comments', ''),
+                'archiveid': '',
+                'archival_remarks': ''
+            }
+        report['photometry'] = {'photometry_group': {}}
+        i = 0
+        for photometry in photometry_list:
+            if photometry.get('brightness'):
+                report_photometry = {
+                    'obsdate': parse_date(photometry.get('date_obs')).strftime('%Y-%m-%d %H:%M:%S'),
+                    'flux': photometry.get('brightness', ''),
+                    'flux_error': photometry.get('brightness_error', ''),
+                    'limiting_flux': photometry.get('limiting_brightness', ''),
+                    'flux_units': convert_flux_units(photometry.get('brightness_unit', 'AB mag')),
+                    'filter_value': str(tns_options.get('filters', {}).get(photometry.get('bandpass'))),
+                    'instrument_value': str(tns_options.get('instruments', {}).get(photometry.get('instrument', ''))),
+                    'exptime': str(photometry.get('exposure_time', '')),
+                    'observer': photometry.get('observer', ''),
+                    'comments': photometry.get('comments', '')
+                }
+                report['photometry']['photometry_group'][str(i)] = report_photometry
+                i += 1
+        at_report[str(len(at_report))] = report
+    return at_report
+
+
+def get_tns_marker():
+    tns_marker = 'tns_marker{"tns_id": "' + str(settings.TNS_CREDENTIALS.get('id')) + '", "type": "bot", "name": "' + settings.TNS_CREDENTIALS.get('name') + '"}'
+    return tns_marker
+
+
+def parse_object_from_tns_response(response_json):
+    at_report_response = response_json.get('data', {}).get('feedback', {}).get('at_report', {})[0]
+    for value in at_report_response.values():
+        if isinstance(value, dict) and 'objname' in value:
+            return value['objname']
+    return None
+
+
+def submit_to_tns(at_report):
+    """ Submit a tns formatted message to the tns server, and returns the TNS object name """
+    data = {'at_report': at_report}
+    payload = {
+        'api_key': settings.TNS_CREDENTIALS.get('api_token'),
+        'data': json.dumps(data, indent=4)
+    }
+    url = urljoin(settings.TNS_BASE_URL, 'api/bulk-report')
+    headers = {'User-Agent': get_tns_marker()}
+    try:
+        response = requests.post(url, headers = headers, data = payload)
+        response.raise_for_status()
+        report_id = response.json()['data']['report_id']
+    except Exception:
+        raise Exception
+
+    attempts = 0
+    object_name = None
+    reply_url = urljoin(settings.TNS_BASE_URL, 'api/bulk-report-reply')
+    reply_data = {'api_key': settings.TNS_CREDENTIALS.get('api_token'), 'report_id': report_id}
+    while attempts < 10:
```

Comment:
This is essentially attempting to "check" the TNS ingestion status 10 times, waiting 1s between each check. If we get an error result, it raises that to the user and don't submit the message to hop. If we don't get a result, then we raise an error that its ingestion is unknown to the user, and also don't submit the message through to hop. Only a good result with the object name in the TNS server results in us submitting the hop message.

Reply:
maybe a breadcrumb comment explaining _why_ we're doing this? (i.e. what is it about TNS behavior that makes this necessary?)